### PR TITLE
Bug fixes and perf improvements for setting providers in AuthUI

### PR DESCRIPTION
**Pull request by SUPERCILEX**
@samtstern We currently throw an exception, but I think we should be smarter about this. Now, everything gets merged together into one unique set of providers and scopes.

### CODE REVIEW

File: auth/src/main/java/com/firebase/ui/auth/AuthUI.java [hunk 1]
```diff
@@ -356,11 +355,33 @@ public void writeToParcel(Parcel parcel, int i) {
...
         }
 
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+
+            IdpConfig config = (IdpConfig) o;
+
+            return mProviderId.equals(config.mProviderId);
```
Comment (SUPERCILEX) [inline auth/src/main/java/com/firebase/ui/auth/AuthUI.java]: @samtstern Identifying providers by provider id only and completely ignoring scopes makes sense to me, do you agree?

File: auth/src/main/java/com/firebase/ui/auth/AuthUI.java [hunk 2]
```diff
@@ -406,7 +427,7 @@ public IdpConfig build() {
     public final class SignInIntentBuilder {
         private int mLogo = NO_LOGO;
         private int mTheme = getDefaultTheme();
-        private LinkedHashSet<IdpConfig> mProviders = new LinkedHashSet<>();
+        private List<IdpConfig> mProviders = new ArrayList<>();
```
Comment (SUPERCILEX) [inline auth/src/main/java/com/firebase/ui/auth/AuthUI.java]: This is much better, we were converting this to a list anyway which seems kinda pointless.

File: auth/src/main/java/com/firebase/ui/auth/AuthUI.java [hunk 3]
```diff
@@ -458,19 +479,30 @@ public SignInIntentBuilder setTosUrl(@Nullable String tosUrl) {
          */
         public SignInIntentBuilder setProviders(@NonNull List<IdpConfig> idpConfigs) {
             mProviders.clear();
-            Set<String> configuredProviders = new HashSet<>();
```
Comment (SUPERCILEX) [inline auth/src/main/java/com/firebase/ui/auth/AuthUI.java]: Less memory usage if you don't overlap providers! 😄

File: auth/src/main/java/com/firebase/ui/auth/AuthUI.java [hunk 4]
```diff
@@ -495,6 +527,15 @@ public SignInIntentBuilder setProviders(@NonNull String... providers) {
             return this;
         }
 
+        private boolean isIdpAlreadyConfigured(@NonNull String providerId) {
```
Comment (SUPERCILEX) [inline auth/src/main/java/com/firebase/ui/auth/AuthUI.java]: I moved it to where it's being used (in the deprecated `setProviders`).

File: auth/src/test/java/com/firebase/ui/auth/AuthUITest.java [hunk 5]
```diff
@@ -63,13 +66,30 @@ public void testCreateStartIntent_shouldHaveEmailAsDefaultProvider() {
...
+        List<IdpConfig> providers = builder.getFlowParams().providerInfo;
+        assertTrue(providers.size() == 2);
+
+        IdpConfig email = providers.get(0);
+        List<String> emailScopes = email.getScopes();
+        assertTrue(email.getProviderId().equals(AuthUI.EMAIL_PROVIDER));
+        assertTrue(emailScopes.size() == 2);
+        assertTrue(emailScopes.contains("A") && emailScopes.contains("B"));
+
+        assertTrue(providers.get(1).getProviderId().equals(AuthUI.GOOGLE_PROVIDER));
```
Comment (SUPERCILEX) [inline auth/src/test/java/com/firebase/ui/auth/AuthUITest.java]: @samtstern Did I get all the possible cases?

**Comment by samtstern** [comment]
@SUPERCILEX what is the use case for specifying the same provider multiple times in the builder?  I think throwing an Exception is good since I doubt the developer meant to specify duplicate providers and we are probably helping them catch a mistake.

**Comment by SUPERCILEX** [comment]
@samtstern If a user enables a setting in the setup of the app, they get logged in with different permissions than the standard setup. This means I have to take my default configuration, remove the `IdpConfig` for Google, and add it back with the new permissions by doing an `addAll` to my existing permissions. Whew! It's kind of a pain, but also not that big of a deal. It would just be easier to add another `IdpConfig` with the extra permission I need and have FirebaseUI merge it for me.

**Comment by samtstern** [comment]
@SUPERCILEX my personal opinion is that there will be more people saved from doing the wrong thing by the error-throwing behavior than there will be people who appreciate the merging of IdpConfig.

**Comment by SUPERCILEX** [comment]
@samtstern Okay, that makes sense. I made some improvements to increase readability and reduce memory thrashing along the way so I'll leave this PR open and focus its changes on that instead.

File: auth/src/main/java/com/firebase/ui/auth/AuthUI.java [hunk 6]
```diff
@@ -515,23 +543,18 @@ public SignInIntentBuilder setAllowNewEmailAccounts(boolean enabled) {
...
-
         public Intent build() {
             return KickoffActivity.createIntent(mApp.getApplicationContext(), getFlowParams());
         }
 
         @VisibleForTesting()
         public FlowParameters getFlowParams() {
+            if (mProviders.isEmpty()) {
+                mProviders.add(new IdpConfig.Builder(EMAIL_PROVIDER).build());
+            }
```
Comment (SUPERCILEX) [inline auth/src/main/java/com/firebase/ui/auth/AuthUI.java]: This is why we should always do this kind of thing in the `build` method: if people call `setProviders(Collections.emptyList())`, we will start the auth flow with no providers. Now, we can guarantee that at least the email provider is provided.

**Comment by samtstern** [comment]
@SUPERCILEX SGTM, we can move forward with the general improvements.

**Comment by samtstern** [comment]
Forgot that you had updated this, but LGTM.  Squashing and merging.

**Comment by SUPERCILEX** [comment]
@samtstern Oops, I forgot to ask you to wait for @amandle's PR to be merged since this one caused a bunch of merge conflicts. I created https://github.com/amandle/FirebaseUI-Android/pull/1 to fix the conflicts. Sorry about that!

**Pull request merged**